src/rpc/blockchain.cpp

```diff
@@ -3171,6 +3097,152 @@ static RPCHelpMan dumptxoutset()
     };
 }
 
+/**
+ * RAII class that creates a temporary database directory in its constructor
+ * and removes it in its destructor.
+ */
+class TemporaryUTXODatabase
+{
+    fs::path m_path;
+public:
+    TemporaryUTXODatabase(const fs::path& path) : m_path(path) {
+        fs::create_directories(m_path);
+    }
+    ~TemporaryUTXODatabase() {
+        try {
+            fs::remove_all(m_path);
+        } catch (...) {
+            LogInfo("Failed to clean up temporary UTXO database at %s, please remove it manually.",
+                    fs::PathToString(m_path));
+        }
+    }
+};
+
+UniValue CreateRolledBackUTXOSnapshot(
+    NodeContext& node,
+    Chainstate& chainstate,
+    const CBlockIndex* target,
+    AutoFile&& afile,
+    const fs::path& path,
+    const fs::path& tmppath)
+{
+    // Create a temporary leveldb to store the UTXO set that is being rolled back
+    std::string temp_db_name{strprintf("temp_utxo_%d", target->nHeight)};
+    fs::path temp_db_path{fsbridge::AbsPathJoin(tmppath.parent_path(), fs::u8path(temp_db_name))};
+    TemporaryUTXODatabase temp_db_cleaner{temp_db_path};
+
+    // Create temporary database
+    DBParams db_params{
+        .path = temp_db_path,
+        .cache_bytes = size_t(1) << 24,  // 16MB cache
+        .memory_only = false,
+        .wipe_data = true,
+        .obfuscate = false,
+        .options = DBOptions{}
+    };
+
+    std::unique_ptr<CCoinsViewDB> temp_db = std::make_unique<CCoinsViewDB>(
+        std::move(db_params),
+        CoinsViewOptions{}
+    );
+
+    LogInfo("Copying current UTXO set to temporary database.");
+    {
+        WITH_LOCK(::cs_main, chainstate.ForceFlushStateToDisk());
+        CCoinsViewCache temp_cache(temp_db.get());
+        temp_cache.SetBestBlock(chainstate.m_chain.Tip()->GetBlockHash());
+
+        std::unique_ptr<CCoinsViewCursor> cursor;
+        WITH_LOCK(::cs_main, cursor = chainstate.CoinsDB().Cursor());
+
+        size_t coins_count = 0;
+        while (cursor->Valid()) {
```

Comment:
Don't we need to hold `cs_main` throughout? What if the utxo set changes while we are copying coins?